### PR TITLE
cmd: fix teardown

### PR DESCRIPTION
**Pull request by Rolinh**
See individual commits for details.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -31,28 +33,59 @@ import (
...
 // This is called by main.main(). It only needs to happen once to the rootCmd.
-func Execute() {
-	fin := maybeProfile()
-	defer fin() // make sure update memory profile is written in the end
-
-	if err := rootCmd.Execute(); err != nil {
+func Execute() error {
+	err := rootCmd.Execute()
+	if err != nil {
 		fmt.Println(err)
```
Comment (glibsm) [inline cmd/root.go]: it's better to return the error and let the caller choose print or not

File: main.go [hunk 2]
```diff
@@ -19,9 +19,13 @@
...
 import (
+	"os"
+
 	"github.com/cilium/hubble/cmd"
 )
 
 func main() {
-	cmd.Execute()
+	if err := cmd.Execute(); err != nil {
+		os.Exit(-1)
```
Comment (glibsm) [inline main.go]: log.Fatal is probably more appropriate with the error message

File: cmd/root.go [hunk 3]
```diff
@@ -31,28 +33,59 @@ import (
...
-	Use:     "hubble",
-	Short:   "CLI",
-	Long:    `Hubble is a utility to observe and inspect recent Cilium routed traffic in a cluster.`,
-	Version: pkg.Version,
+	Use:          "hubble",
+	Short:        "CLI",
+	Long:         `Hubble is a utility to observe and inspect recent Cilium routed traffic in a cluster.`,
+	SilenceUsage: true,
+	Version:      pkg.Version,
+	PersistentPreRunE: func(_ *cobra.Command, _ []string) error {
```
Comment (glibsm) [inline cmd/root.go]: maybe move the inners to a separate function - the command definition will the remain cleaner

File: cmd/serve/serve.go [hunk 4]
```diff
@@ -101,12 +102,14 @@ func New(log *logrus.Entry) *cobra.Command {
 				int(eventQueueSize),
 				log,
 			)
+			ctx, cancel := context.WithCancel(context.Background())
+			setupSigHandler(ctx, cancel)
```
Comment (glibsm) [inline cmd/serve/serve.go]: i don't think cancel is called in all cases with this setup (crashing or other error without ctrl-c)

File: .gitignore [hunk 5]
```diff
@@ -1,2 +1,3 @@
 /hubble
 install/kubernetes/hubble.yaml
+*.pprof
```
Comment (glibsm) [inline .gitignore]: we don't need to remove it from here, but this is a fantastic addition to your global gitignore so pprof files are never checked in across all projects

File: cmd/serve/serve.go [hunk 6]
```diff
@@ -243,23 +243,24 @@ func setupListeners(listenClientUrls []string) (listeners map[string]net.Listene
...
-				"Shutting down after the configured duration",
-			)
-			os.Exit(0)
-		}()
+		go func(ctx context.Context) {
+			select {
+			case <-time.After(serveDuration):
+				log.WithField("duration", serveDuration).Info(
+					"Shutting down after the configured duration",
+				)
```
Comment (Rolinh) [inline cmd/serve/serve.go]: Just realized that the context should be cancelled here. Do we want to keep the server shutdown after duration feature though? What's the use-case for this?

Comment (gandro) [reply cmd/serve/serve.go]: I believe the `serveDuration` was introduced to make integration tests simpler, but the integration tests never landed. However, given that we now have embedded Hubble in Cilium, this the `serveDuration` is becoming less and less useful, as even if we stop the Hubble server, `cilium-agent` will continue to run.

I agree with @michi-covalent (https://github.com/cilium/hubble/pull/182#pullrequestreview-371363840) that we should simply remove `serveDuration`.

Comment (Rolinh) [reply cmd/serve/serve.go]: OK, I'll take care of this.

**Comment by Rolinh** [comment]
> i kind of want to just get rid of serveDuration. i don't think anybody's using it.

See https://github.com/cilium/hubble/pull/185

**Pull request merged**